docs/manpages/tracker-control.1
.TH tracker-control 1 "September 2009" GNU "User Commands"

.SH NAME
tracker-control \- Manage Tracker processes and data

.SH SYNOPSIS
\fBtracker-control\fR [\fIOPTION\fR...]

.SH DESCRIPTION
.B tracker-control
manages and checks status of all Tracker processes and data.

Controls Tracker both at process level, and at entity level (store, miners).

To start or stop miners, you can use
.B \-\-start.
The store is started automatically by the D-Bus calls from the miners.

It also allows checking the status of the Tracker store and all data miners.
For
.B tracker-store
, the status is always
.B Idle
unless it is restoring a backup and/or replaying a journal (regardless of
load from applications or miners). For a list of common statuses, see
.B \-\-list\-common\-statuses.

The miners can be paused or resumed using this command and you can
also list miners running and available.

.SH COMMON OPTIONS
.TP
.B \-?, \-\-help
Show summary of options by group.
.TP
.B \-?, \-\-help-all
Show all groups and options.
.TP
.B \-V, \-\-version
Returns the version of this program.

.SH GENERAL OPTIONS
.TP
.B \-p, \-\-list\-processes
This lists all Tracker processes in the system.
.TP
.B \-k, \-\-kill=[all|store|miners]
This uses SIGKILL to stop all Tracker processes found matching the
parameter, if no extra parameter is passed,
.B all
will be assumed. This is not advised unless you are having problems
stopping Tracker in the first place. This
.B GUARANTEES
death.
.TP
.B \-t, \-\-terminate=[all|store|miners]
This uses SIGTERM to stop all Tracker processes found matching the
parameter, if no extra parameter is passed,
.B all
will be assumed. This is recommended over \-\-kill because it gives
the processes time to shutdown cleanly.
.TP
.B \-r, \-\-hard-reset
This kills all processes in the same way that
.B \-\-kill
does but it also removes all databases. Restarting
.B tracker-store
re-creates the databases.
.TP
.B \-e, \-\-soft-reset
A soft reset works exactly the same way that
.B \-\-hard-reset
does, with the exception that the backup and journal are not removed.
These are restored when
.B tracker-store
is restarted.
.TP
.B \-c, \-\-remove-config
This removes all config files in $HOME/.config/tracker. All files
listed are files which were found and successfully removed.
Restarting the respective processes re-creates the default
configuration files.
.TP
.B \-s, \-\-start
Starts all miners. This indirectly starts tracker-store too because it
is needed for miners to operate properly.
.TP
.B \-b, \-\-backup=FILE
Begins backing up the Tracker databases to the
.B FILE
given.
.TP
.B \-o, \-\-restore=FILE
Begins restoring a previous backup (see
.B \-\-backup
) to the Tracker databases. The
.B FILE
points to the location of the backup.
.TP
.B \-\-get-log-verbosity
This displays the log verbosity for ALL components using GSettings for
this configuration. For possible values, see
.B \-\-set-log-verbosity.
.TP
.B \-\-set-log-verbosity=[debug|detailed|minimal|errors]
This sets the log verbosity for ALL components using GSettings using
this configuration option ('verbosity').
.TP
.B \-\-collect-debug-info
Useful when debugging problems to diagnose the state of Tracker on
your system. The data is output to stdout. Useful if bugs are filed
against the project itself.

Data collected includes Tracker version in use, disk space available,
size of the databases on the disk, the configuration in use, states of
the index (e.g. last filesystem crawl, data set locale, etc.) and
finally statistics about the data in the database (e.g. how many
nfo:FileDataObject resources exist).

.SH STATUS OPTIONS
.TP
.B \-S, \-\-status
Show the current status of all Tracker entities (store and all available
miners).
.TP
.B \-F, \-\-follow
Follow status changes as they happen. This requires Ctrl+C to stop and
return to the command line. Each new status is put on a new line.
.TP
.B \-w, \-\-watch=[ONTOLOGY]
Watch changes that happen to the database in real time. This requires
Ctrl+C to stop and return to the command line.

If the
.B ONTOLOGY
is unspecified, all changes are shown. The
.B ONTOLOGY
can be a comma separated list of shorthand or long hand ontology
properties. For example:

.nf
    $ tracker-control -w nie:url,nie:mimeType,nfo:fileSize,nie:dataSource
    Now listening for resource updates to the database
    All nie:plainTextContent properties are omitted

    Press Ctrl+C to stop

    'nfo:Document'
       'nfo:fileSize' = '1770'
       'nie:dataSource' = 'http://www.tracker-project.org/ontologies/tracker#extractor-data-source'
       'nie:mimeType' = 'text/plain'
       'nie:url' = 'file:///home/martyn/.bash_aliases'
    'nfo:Document'
       'nie:dataSource' = 'http://www.tracker-project.org/ontologies/tracker#extractor-data-source'

    ...
.fi

.TP
.B \-\-list-common-statuses
This will list statuses most commonly produced by miners and the
store. These statuses are not translated when sent over D-Bus and
should be translated by each application. These are not considered
static and are subject to change at any point.

Additionally, these statuses are not the only ones which may be
reported by a miner. There may be other states pertaining to the
specific roles of the miner in question.

.SH MINER OPTIONS
.TP
.B \-m, \-\-reindex-mime-type=MIME
Re-index files which match the \fIMIME\fR type supplied. This is
usually used when installing new extractors which support \fIMIME\fR
types previously unsupported. This forces Tracker to re-index those
files. You can use
.B \-\-reindex-mime-type
more than once per \fIMIME\fR type.
.TP
.B \-f, \-\-index-file=FILE
(Re)index a file matching the \fIFILE\fR type supplied.
.TP
.B \-\-list-miners-running
This will list all miners which have responded to a D-Bus call.
Sometimes it is helpfult to use this command with
.B \-\-list-miners-available.
.TP
.B \-\-list-miners-available
This will list all miners which are available even if they are not
running at the moment.
.TP
.B \-\-pause-details
For listing all miners which are paused and the reasons for being
paused, you can use this. It will also display the application that
requested the pause too.
.TP
.B \-\-miner=MINER
This argument is used with
.B \-\-pause
or
.B \-\-resume
to say which miner you want to pause or resume. You can use the full
D-Bus name, e.g.
.B org.freedesktop.Tracker1.Miner.Files
OR you can use the suffix, e.g.
.B Files
.TP
.B \-\-pause=REASON
The REASON here is useful to know WHY the miner should be paused. A
miner can be paused many times by multiple applications. Only when all
pauses have been resumed will it continue. If successful, a cookie
will be given to uniquely identify the request. This cookie is used to
resume the pause at a later stage.
.TP
.B \-\-pause-for-process=REASON
This works exactly the same way as
.B \-\-pause
with the exception that it only keeps the pause active while the
calling process is alive. As soon as you press Ctrl+C the pause is
resumed automatically.
.TP
.B \-\-resume=COOKIE
The COOKIE is given by a successful
.B \-\-pause
command. It is a number which identifies each pause request. When all
pauses have been resumed, the miner will resume working.

.SH ENVIRONMENT
.TP
.B TRACKER_MINERS_DIR
Miners that are installed should have a .desktop file in
/usr/share/tracker/miners/, this is how tracker-control knows about
miners which exist but may not have not been started. This environment
variable allows the location that miner .desktop files are loaded from
to be overridden by setting it to a new path.
.TP
.B TRACKER_MINERS_DIR_DISABLED
If you don't have .desktop files for miners, you can use this
environment variable to get status information from \fBRUNNING\fR
processes. This doesn't give any indication for miners \fBNOT\fR
running. This uses registered miners on DBus to know about existing
Tracker based data miners.

.SH SEE ALSO
.BR tracker-store (1).


NEWS
NEW in 1.1.1 - 2014-07-10
=========================

The main changes since 1.1.0 are:

  * build: Use major version only PkgConfig files. Previously, we had
    $library-$major-$minor (e.g. tracker-sparql-1.2, when it should be
    tracker-sparql-1.0) to avoid creating unnecessary work for
    developers using Tracker. This release corrects this and it was
    planned originally.

NEW in 1.1.0 - 2014-07-10
=========================

The main changes since 1.0.0 are:

  * build: Added CC_CHECK_FLAGS_APPEND m4 macros from xinelib/systemd projects to check GCC compiler flags we can use
  * tracker-miner-apps: New process to mine data about applications, previously part of tracker-miner-fs. Can be disabled with --disable-miner-apps
  * tracker-miner-user-guides: New process to mine data about user guides and help content, previously part of tracker-miner-fs. Can be disabled with --disable-user-guides
  * tracker-extract: [BMP] Added an new extractor
  * tracker-extract: [ISO] Extract language, copyright and author information about ISO images using new ontology
  * tracker-extract: [ODT] Improved extraction of content so we don't stop on embedded tabs and line breaks
  * tracker-control: Added a lot of strings previsouly untranslated
  * tracker-control: Removed shorthand command line switches for less common arguments (--list-*, --pause-details)
  * tracker-control: Added --watch command line option to watch changes to the database.
  * tracker-info: Add --resource-is-iri or -i to avoid filename to URN lookups
  * tracker-info: Don't show "(null)" for nie:InformationElements with no nie:url
  * tracker-preferences: General UI clean up and modernization fixing many deprecations where old GTK+ widgets were used
  * tracker-preferences: Added 'control' tab to allow users to enable finer control of what's indexed (e.g. numbers - which are normally disabled)
  * tracker-preferences: Add AppData with screenshots for app stores
  * tracker-needle: Add AppData with screenshots for app stores

Translations:

  * Updated ca: Gil Forcada, Pau Iranzo
  * Updated cs: Marek Černocký
  * Updated de: Christian Kirbach, Benjamin Steinwender
  * Updated el: Dimitris Spingos, marablack3, Dimitris Spingos (Δημήτρης Σπίγγος)
  * Updated es: Daniel Mustieles
  * Updated eu: Inaki Larranaga Murgoitio, Iñaki Larrañaga Murgoitio
  * Updated gl: Fran Diéguez, Fran Dieguez
  * Updated it: Milo Casagrande
  * Updated ko: Seong-ho Cho
  * Updated lt: Aurimas Černius
  * Updated pl: Piotr Drąg
  * Updated ru: Yuri Myasoedov
  * Updated sr: Мирослав Николић
  * Updated sr@latin: Мирослав Николић, Miroslav Nikolić
  * Updated tr: Muhammet Kara
  * Updated uk: Daniel Korostil
  * Updated zh_CN: YunQiang Su, Yunqiang Su


NEW in 1.0.0 - 2014-03-24
=========================

The main changes since 0.16.x are:

  * build: Require libmediaart >= 0.1.0 (soft dependency), code has been exported from Tracker to this library
  * build: Require libicu >= 4.8.1.1, first version with pkg-config support, no version specified before
  * build: Require Vala >= 0.18.0 (was 0.16.0) to avoid Vala 0.16 g_dbus_message_get_body() binding bug
  * build: Require GLib >= 2.38 (was 2.35.1) after removing all deprecated g_test_trap_fork() calls for _subprocess()
  * build: Require libosinfo >= 0.2.9 (was 0.0.2)
  * build: Updated GSF dependency to >=1.14.24
  * build: Improve SQLite FTS feature checking for cross compilation
  * build: Improve description of build switches
  * build: Remove ALL GLib marshallers. We depend on GLib (>= 2.30) which depends on libffi enabling us to do this
  * build: Make sure --target-glib used with Vala is the correct GLib (was 2.34, should be $GLIB_REQUIRED), removes Vala g_type_init() warnings.
  * build: Do not maintain LT versioning for internal libraries libtracker-{common,data,extract}
  * build: GTester Makefile rules have been updated
  * build: Improve portability, avoid some GNUisms and #ifdef linux specific API
  * build: Improved portabiliy for OpenBSD and Solaris
  * build: Use UPower >=0.99 if available
  * all: Make GBusType run-time configurable with TRACKER_BUS_TYPE environment variable
  * New Feature: Passive extraction: Specific metadata extraction (e.g. image orientation) is done by listening for insert signals for resource metadata (e.g. file name, size, etc) and done passively rather than being required before inserting ANY information about a resource. This officially makes tracker-extract a miner of special sorts
  * New Feature: SPARQL functions tracker:normalize() and tracker:unaccent()
  * New Feature: libav extractor for extracting audio/video metadata
  * New Feature: tracker-store GraphUpdated signal delay now configurable: This signal is what's used to notify of resource changes in the database (e.g. new file added), previously it was always 1 second, now it can be configured
  * libtracker-common: Added strnlen() compatibility function
  * libtracker-data: Fix AS ?foo handling in FTS queries
  * libtracker-direct: Don't allow connection if current and db locale mismatch
  * libtracker-miner: Major API cleanups,
    - TrackerMinerWeb has been removed, replaced by TrackerMinerOnline
    - TrackerMinerManager has been splitted into a separate libtracker-control
    - TrackerCrawler, TrackerStorage and the thumbnailer/mediaart helper functions are no longer public.
  * libtracker-extract: This library has been made private. Implementing a TrackerDecorator is the recommended way to provide additional data.
  * libtracker-control: New library, containing TrackerMinerManager, it requires using tracker-control-1.0.pc
  * tracker-miner-fs: optional prioritization of folders has been reimplemented, used on removable volumes and DBus requests
  * tracker-extract: Prioritization of rdf:types has been implemented, so certain file types are extracted before others, on demand of applications.

src/tracker-control/tracker-control.c
/*
 * Copyright (C) 2008, Nokia <ivan.frade@nokia.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "config.h"

#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <locale.h>
#include <errno.h>
#include <sys/types.h>
#include <signal.h>

#include <glib.h>
#include <glib/gi18n.h>
#include <glib/gprintf.h>
#include <glib/gstdio.h>
#include <gio/gio.h>

#include <libtracker-common/tracker-common.h>

#include "tracker-control.h"

#define ABOUT	  \
	"Tracker " PACKAGE_VERSION "\n"

#define LICENSE	  \
	"This program is free software and comes without any warranty.\n" \
	"It is licensed under version 2 or later of the General Public " \
	"License which can be viewed at:\n" \
	"\n" \
	"  http://www.gnu.org/licenses/gpl.txt\n"

static gboolean print_version;

static GOptionEntry common_entries[] = {
	{ "version", 'V', 0, G_OPTION_ARG_NONE, &print_version,
	  N_("Print version"),
	  NULL },
	{ NULL }
};

gboolean
tracker_control_dbus_get_connection (const gchar      *name,
                                     const gchar      *object_path,
                                     const gchar      *interface_name,
                                     GDBusProxyFlags   flags,
                                     GDBusConnection **connection,
                                     GDBusProxy      **proxy)
{
	GError *error = NULL;

	*connection = g_bus_get_sync (TRACKER_IPC_BUS, NULL, &error);

	if (!*connection) {
		g_critical ("%s, %s",
		            _("Could not get D-Bus connection"),
		            error ? error->message : _("No error given"));
		g_clear_error (&error);

		return FALSE;
	}

	*proxy = g_dbus_proxy_new_sync (*connection,
	                                flags,
	                                NULL,
	                                name,
	                                object_path,
	                                interface_name,
	                                NULL,
	                                &error);

	if (error) {
		g_critical ("%s, %s",
		            _("Could not create D-Bus proxy to tracker-store"),
		            error ? error->message : _("No error given"));
		g_clear_error (&error);

		return FALSE;
	}

	return TRUE;
}

int
main (int argc, char **argv)
{
	GOptionContext *context;

	setlocale (LC_ALL, "");

	bindtextdomain (GETTEXT_PACKAGE, LOCALEDIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);

	/* Translators: this messagge will apper immediately after the  */
	/* usage string - Usage: COMMAND [OPTION]... <THIS_MESSAGE>     */
	context = g_option_context_new (_(" - Manage Tracker processes and data"));

	/* Groups */
	g_option_context_add_group (context,
	                            tracker_control_general_get_option_group ());
	g_option_context_add_group (context,
	                            tracker_control_status_get_option_group ());
	g_option_context_add_group (context,
	                            tracker_control_miners_get_option_group ());
	/* Common options */
	g_option_context_add_main_entries (context, common_entries, NULL);

	g_option_context_parse (context, &argc, &argv, NULL);
	g_option_context_free (context);

	if (print_version) {
		g_print ("\n" ABOUT "\n" LICENSE "\n");
		return EXIT_SUCCESS;
	}

	/* General options? */
	if (tracker_control_general_options_enabled ()) {
		if (tracker_control_status_options_enabled ()) {
			g_printerr ("%s\n",
			            _("General and Status options cannot be used together"));
			return EXIT_FAILURE;
		}

		if (tracker_control_miners_options_enabled ()) {
			g_printerr ("%s\n",
			            _("General and Miners options cannot be used together"));
			return EXIT_FAILURE;
		}

		return tracker_control_general_run ();
	}

	/* Status options? */
	if (tracker_control_status_options_enabled ()) {
		if (tracker_control_miners_options_enabled ()) {
			g_printerr ("%s\n",
			            _("Status and Miners options cannot be used together"));
			return EXIT_FAILURE;
		}

		return tracker_control_status_run ();
	}

	/* Miners options? */
	if (tracker_control_miners_options_enabled ()) {
		return tracker_control_miners_run ();
	}

	/* Unknown options? */
	if (argc > 1) {
		gint i = 1;

		g_printerr ("%s: ",
		            _("Unrecognized options"));
		for (i = 1; i < argc; i++) {
			g_printerr ("'%s'%s",
			            argv[i],
			            i == (argc - 1) ? "\n" : ", ");
		}
		return EXIT_FAILURE;
	}

	/* No-args output */
	tracker_control_general_run_default ();
	printf ("\n");

	tracker_control_status_run_default ();
	printf ("\n");

	tracker_control_miners_run_default ();

	return EXIT_SUCCESS;
}
